server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -2328,22 +2343,32 @@ public void onFailure(final Exception e) {
                     failShard("exception during primary term transition", e);
                 } catch (AlreadyClosedException ace) {
                     // ignore, shard is already closed
+                } finally {
+                    listener.onFailure(e);
                 }
             }
 
             @Override
             public void onResponse(final Releasable releasable) {
-                try (Releasable ignored = releasable) {
+                final RunOnce releaseOnce = new RunOnce(releasable::close);
+                boolean success = false;
+                try {
                     assert operationPrimaryTerm <= pendingPrimaryTerm;
                     termUpdated.await();
                     // indexShardOperationPermits doesn't guarantee that async submissions are executed
                     // in the order submitted. We need to guard against another term bump
                     if (operationPrimaryTerm < newPrimaryTerm) {
                         operationPrimaryTerm = newPrimaryTerm;
-                        onBlocked.run();
+                        listener.onPrimaryTermUpdate();
                     }
+                    listener.onResponse(releaseOnce::run);
+                    success = true;
                 } catch (final Exception e) {
                     onFailure(e);
+                } finally {
+                    if (success == false) {
+                        releaseOnce.run();
```

Comment:
It was mostly a safeguard in case the listener throws an exception and didn't close the release itself.

> I would prefer to keep the semantics like that.

+1